cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/SparkPipelineRunner.java

```diff
@@ -898,6 +900,25 @@ private boolean shouldCache(Dag dag, String stageName, Set<String> branchers, Se
     return !Sets.intersection(dag.getSources(), parents).isEmpty();
   }
 
+  /**
+   * Wraps a Spark Collection with RecordInfo for the stage.
+   *
+   * @param stageName name of the stage to use.
+   * @param collection Collection to use.
+   * @return Instance of a spark collection with RecordInfo attached to output records.
+   */
+  private SparkCollection<RecordInfo<Object>> mapToRecordInfoCollection(String stageName,
+                                                                        SparkCollection<Object> collection) {
+    // For SQLEngineCollection or WrappedSparkCollection, we wrap the collection in order to not force a
+    // premature/unnecessary pull operation from the SQL engine.
+    if (collection instanceof SQLBackedCollection) {
+      return new WrappedSQLEngineCollection<>((SQLBackedCollection<Object>) collection,
```

Comment:
this is kind of brittle in that it delays the first operation, but subsequent ones still pull the data. So if we were to map again for some reason in the future, it could have unintended consequences.

It would also be kind of hacky, but wondering if the SQLEngineCollection should map the RDD to a RDD<RecordInfo> when it pulls, and omit the map to RecordInfo when the sql engine is used.

Reply:
I'm actually delaying 2 operations:

1) Map to RecordInfo
2) Map back again to record (which is used when invoking the addEmitted method).

Just delaying these 2 operations saves me from having to do a pull in the scenario where multiple joins are consecutive. Since the records are already in the SQL engine, successive join operations will not never really operate on the underlying dataset, thus sparing me from needing to do a pull.

I considered refactoring the addEmitted method to support collections without recordInfo when the collection is a SQL engine collection, but it has 2 side effects:

1) this means I need to skip the logic that we use to filter output records based on the output port for a given record. While this is not a problem for joins, it might become a problem as we add additional functionality in the SQL engine.
2) It's one more code path we need to maintain and remember to update.

I took the "lazy" (in implementation and in execution) approach of just delaying `pull` operations until I have no other option (such as a wrangler step or a sink, for example).